src/common/set.c

```diff
@@ -2706,58 +2722,30 @@ util_poolset_append_new_part(struct pool_set *set, size_t size)
 
 	struct pool_set_directory *d;
 	size_t directory_id;
+	char *path;
+	size_t path_len;
 
-	struct part {
-		char *path;
-		size_t path_len;
-		int fd;
-	} *parts = Zalloc(sizeof(*parts) * set->nreplicas);
-	if (parts == NULL)
-		return -1;
-
-	for (unsigned r = 0; r < set->nreplicas; ++r) {
-		if (util_replica_reserve(&set->replica[r],
-			set->replica[r]->nparts + 1) != 0)
-			goto err_part_init;
-
+	unsigned r;
+	for (r = 0; r < set->nreplicas; ++r) {
 		struct pool_replica *rep = set->replica[r];
 
 		directory_id = set->next_directory_id %
 			VEC_SIZE(&rep->directory);
 		d = VEC_GET(&rep->directory, directory_id);
 
-		struct part *p = &parts[r];
-
-		p->path_len = strlen(d->path) + PMEM_FILE_MAX_LEN;
-		if ((p->path = Malloc(p->path_len)) == NULL)
+		path_len = strlen(d->path) + PMEM_FILE_MAX_LEN;
+		if ((path = Malloc(path_len)) == NULL) {
+			ERR("!Malloc");
 			goto err_part_init;
+		}
 
-		snprintf(p->path, p->path_len, "%s/%0*u%s",
+		snprintf(path, path_len, "%s" OS_DIR_SEP_STR "%0*u%s",
 			d->path, PMEM_FILE_PADDING, set->next_id, PMEM_EXT);
 
-		p->fd = os_open(p->path, O_RDWR | O_CREAT | O_EXCL,
-			S_IRUSR | S_IWUSR);
-		if (p->fd < 0)
-			goto err_part_init;
-
-		if (os_posix_fallocate(p->fd, 0, (os_off_t)size) != 0)
-			goto err_part_init;
-	}
-
-	/* we cannot easily recover during this loop */
-	for (unsigned r = 0; r < set->nreplicas; ++r) {
-		struct part *p = &parts[r];
-
-		/* the capacity is reserved, so this should never fail */
-		if (util_replica_add_part(&set->replica[r], p->path, size) != 0)
+		if (util_replica_add_part(&set->replica[r], path, size) != 0)
```

Comment:
this could fail now, but since you are not allocating files we can gracefully handle the error.